### ISSUE TITLE
Error matching generic to `__wrapped__` protocol?

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#9399]
**Describe the bug**
Passing a class with a generic `__wrapped__` attribute raises a strange error, non-generic is fine.  This might be as intended, I'm not sure if there's something special about `__wrapped__`.

**Code or Screenshots**

```python
from typing import Protocol, Any
from collections.abc import Callable

class Wrapps[**P, R](Protocol):
    __wrapped__: Callable[P, R]
class WrappsAny(Protocol):
    __wrapped__: Any
class WrappsT[T](Protocol):
    __wrapped__: T

class AnyFnWrapper[C: Callable[..., Any]]:
    __wrapped__: C
class PFnWrapper[**P, R]:
    __wrapped__: Callable[P, R]

def foo[**P, R](arg: Wrapps[P, R]):
    ...
def foot[T](arg: WrappsT[T]):
    ...
def fooany(arg: WrappsAny):
    ...

def bar[**P, R](arg: AnyFnWrapper[Callable[P, R]]):
    # Argument of type "AnyFnWrapper[(**P@bar) -> R@bar]" cannot be assigned to parameter "arg" of type "Wrapps[P@foo, R@foo]" in function "foo"
    #   "AnyFnWrapper[(**P@bar) -> R@bar]" is incompatible with protocol "Wrapps[P@foo, R@foo]"
    #     Could not bind method "<anonymous>" because "AnyFnWrapper[(**P@bar) -> R@bar]" is not assignable to parameter "args"
    #       Type "AnyFnWrapper[(**P@bar) -> R@bar]" is not assignable to type "P@bar.args"  (reportArgumentType)
    foo(arg)
    foot(arg)
    fooany(arg)

def baz[**P, R](arg: PFnWrapper[P, R]):
    # similar error
    foo(arg)
    foot(arg)
    fooany(arg)

def works[P: PFnWrapper[..., Any]](arg: P) -> P:
    # no error
    foo(arg)
    foot(arg)
    fooany(arg)
    return arg

def works2(arg: PFnWrapper[..., Any]):
    # no error
    foo(arg)
    foot(arg)
    fooany(arg)
```

**VS Code extension or command-line**
cli/playground/latest

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Here are the (rather lengthy) details...

In your protocol definitions, you're defining `__wrapper__` to be an instance variable. The [typing spec](https://typing.readthedocs.io/en/latest/spec/protocol.html#protocol-members) indicates that a variable declared within a protocol class body that does not use the `ClassVar` type qualifier should be treated as an instance variable.

The typing spec is unfortunately not as clear about how type checkers should interpret class-scoped variables within regular (non-protocol) classes. You can read about pyright's approach [here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=class-and-instance-variables). I think pyright's approach is defensible, but there are alternative ways that one could interpret annotated class-scoped attributes.

In the concrete classes `AnyFnWrapper` and `PFnWrapper`, pyright interprets `__wrapper__` as a [regular class variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=regular-class-variables).

When a callable object is stored in a class variable, a reference to that variable through an object instance causes the callable to be bound to the object instance. Consider this example:

```python
# Note that this function has a "self" parameter
def my_callable(self: "A"):
    return None

class A:
    x: Callable[..., None] = my_callable

a = A()
a.x() # OK because self argument is provided implicitly through binding
```

I can't tell from your code whether you intend for `__wrapped__` to be a pure instance variable or a regular class variable.

If you change your class definition for `AnyFnWrapper` to the following, pyright will assume that `__wrapped__` is a [pure instance variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=pure-instance-variables), and this will eliminate the type errors you're seeing in your code.

```python
class AnyFnWrapper[C: Callable[..., Any]]:
    def __init__(self):
        self.__wrapped__: C
```

In this case, `__wrapped__` is guaranteed not to be a class variable — an assumption that pyright enforces. Because it's guaranteed not to be a class variable, binding is never performed — that is, the caller doesn't receive an implicit `self` argument.

As I mentioned, the [typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) currently provides little or no guidance to type checker authors about how to distinguish between class variables that can be overwritten by instances and those that cannot.

I recently gave [a talk](https://docs.google.com/presentation/d/1ash1ellodiV6FTyVnY-Q6V2G5vJ68ycWBr6MYkayC1U/edit#slide=id.g2770f948faa_0_5) where I listed some priorities for areas I'd like to see clarified in the typing spec. You can see that "Instance and class variables" is on this list.

I don't plan to make any changes to pyright's current behavior in this regard unless/until the typing spec is updated to specify that it should work differently. In the absence of a specified behavior, I think pyright's behavior is defensible, and changing it would be disruptive to existing pyright users.

**Comment by alwaysmpe** [comment]
Thank you for the explanation, I'll have a watch of your talk

**Comment by alwaysmpe** [comment]
And your comment has led me to the workaround of `Concatenate[Self, P]`, thanks, just need to be careful mixing protocols and classes
